plugin/caddyfileprocessor.go

```diff
@@ -0,0 +1,105 @@
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"sort"
+	"strings"
+
+	"github.com/caddyserver/caddy"
+	"github.com/caddyserver/caddy/caddyfile"
+
+	_ "github.com/caddyserver/caddy/caddyhttp" // plug in the HTTP server type
+)
+
+// ProcessCaddyfile validate and removes wrong server blocks from caddyfile
+func ProcessCaddyfile(content string) string {
+	validDirectives := caddy.ValidDirectives("http")
+
+	serverBlocks, _ := caddyfile.Parse("", bytes.NewReader([]byte(content)), validDirectives)
+
+	var validServerBlocks []caddyfile.ServerBlock
+
+	for _, serverBlock := range serverBlocks {
+		subContent := stringifyServerBlock(&serverBlock)
+
+		newInput := caddy.CaddyfileInput{
+			ServerTypeName: "http",
+			Contents:       []byte(subContent),
+		}
+
+		err := caddy.ValidateAndExecuteDirectives(newInput, nil, true)
+		if err == nil {
+			validServerBlocks = append(validServerBlocks, serverBlock)
+		} else {
+			log.Printf("[WARN] Removing invalid server block: %v\n%v\n", err, subContent)
+		}
+	}
+	return stringifyServerBlocks(validServerBlocks)
+}
+
+func stringifyServerBlocks(serverBlocks []caddyfile.ServerBlock) string {
+	var writer bytes.Buffer
+
+	for _, serverBlock := range serverBlocks {
+		writeServerBlock(&writer, &serverBlock)
+	}
+
+	return writer.String()
+}
+
+func stringifyServerBlock(serverBlock *caddyfile.ServerBlock) string {
+	var writer bytes.Buffer
+	writeServerBlock(&writer, serverBlock)
+	return writer.String()
+}
+
+func writeServerBlock(writer *bytes.Buffer, serverBlock *caddyfile.ServerBlock) {
+	writer.WriteString(fmt.Sprintf("%v {\n", strings.Join(serverBlock.Keys, " ")))
+
+	for _, directiveName := range getSortedDirectiveNames(serverBlock.Tokens) {
+		indent := 1
+		newLine := true
+		tokenLine := -1
+		for _, token := range serverBlock.Tokens[directiveName] {
+			if token.Text == "}" {
+				indent--
+			}
+			if token.Line != tokenLine {
```

Comment:
@mholt
Strange that caddyfile.Parse doesn't return token for line breaks inside directives. So I need to rely on the original token line to recognize line breaks.

```
proxy / service2:5000/api {
  except /a /b
  transparent
}
```
Return tokens:
```
["proxy", "/", "service2:5000/api", "{", "/a", "/b", "transparent", "}"]
```
Or something like that

Reply:
Yeah, the v2 parser (already committed to Caddy's v2 branch) cleans this up a bit by introducing something called "segments" which were always really there implicitly. A segment starts with a directive and goes until the end of the line or the end of the block that line opens. That's what is happening here.

You will find working with tokens is much easier if you use a Dispenser (in the caddyfile package). Same goes for v2 when that lands.